design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,398 @@
+```yaml
+  conditions:
+  - type: Available
+    status: False
+    reason: UnsyncronizedIPPool
+    message: Failed re constructing the IP pool, it may be not in sync with nodes.
+```
+
+The modified/created IP pool has overlaps on the range with another one.
+
+```yaml
+  conditions:
+  - types: Available
+    status: False
+    reason: OverlapingIPPool
+    message: The NodeIPPool XYZ is overlapping with this one.
+```
+
+To assign an IP address from a pool to an interface using NNCP, a special
+capture syntax is introduced understood by kubernetes-nmstate
+`ippool.[name].allocate` that will trigger the ip allocation mechanism.
+It will inject the result on the captured state passed to nmpolicy so it can be
+consumed by it, the `ippool.[name].allocate` tag will return a struct with the
+following fields `ip` and `prefix-length`.
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: static-ip
+capture:
+    address-from-traffic-1: ippool.traffic-1.allocate
+    address-from-traffic-2: ippool.traffic-2.allocate
+desiredState:
+    interfaces:
+    - name: eth1
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-1.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-1.prefix-length }}"
+        dhcp: false
+        enabled: true
+    - name: eth2
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-2.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
+        dhcp: false
+        enabled: true
+```
+
+The allocated offsets will be marked at the `NodeNetworkConfigurationEnactment`
+status and will be used to re-construct the nmstate on reboot and to discover
+if new allocation if needed or to de-allocate them the pool is no longer
+referenced, this is done on distributed fashion by each handler at their nodes.
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeNetworkConfigurationEnactment
+metadata:
+  name: static-ip.node01
+status:
+  ippool:
+    traffic-1:
+      allocated-offsets: [6]
+    traffic-2:
+      allocated-offsets: [8]
+```
+
+### API Extensions baremetal IPI
+
+At baremetal IPI the Nmstate configuration is passed to openshift installer
+using the baremetal host `networkConfig` field [1].
+
+[1] https://github.com/openshift/installer/blob/ca8aacd3c931eefabfddafb49cb44b76f24653a6/docs/user/metal/install_ipi.md#describing-hosts
+
+To accommodate the ippools, capture and nmstate new fields will be needed there
+the following is an example of the install-config.yaml from openshift with
+the new fields:
+
+```yaml
+apiVersion: v1
+baseDomain: test.metalkube.org
+metadata:
+  name: ostest
+platform:
+  baremetal:
+    networkPolicy:
+      ipPools:
+        traffic-1:
+          cidr: 10.10.10.0/24
+          start: 1
+          end: 10
+          exclude: [4, 7]
```

Comment:
are these indexes ?... or just the last byte of the IP in decimal format ? Because they look like the latter. 

How will this fit a CIDR with - let's say - 16 bits for the network ?

Reply:
This example is wrong since `NodeIPPool` was changed since and now start, end and exclude use full IP address, only the allocate-offsets use index since those will be a massive number of entries (we can have cluster with thousand nodes).